### PR TITLE
Ensure packages can be built with TypeScript strict flag

### PR DESCRIPTION
**Pull request by devversion**
Implemented as part of FW-1265. See individual commits for description of changes.

### CODE REVIEW

File: packages/compiler/src/render3/view/template.ts [hunk 1]
```diff
@@ -1609,7 +1611,7 @@ export class BindingScope implements LocalResolver {
   }
 
   generateSharedContextVar(retrievalLevel: number) {
-    const lhs = o.variable(CONTEXT_NAME + this.freshReferenceName());
+    const lhs: o.ReadVarExpr = o.variable(CONTEXT_NAME + this.freshReferenceName());
```
Comment (gkalpak) [inline packages/compiler/src/render3/view/template.ts]: Why this change (since `o.variable` [already returns][1] `o.ReadVarExpr`)?

[1]: https://github.com/angular/angular/blob/c4ab32f664ee8f4619893db508fc2f0cf0f132f3/packages/compiler/src/output/output_ast.ts#L1490-L1493

File: packages/tsconfig-build-strict.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "./tsconfig-build.json",
+  "compilerOptions": {
+    "strict": true
```
Comment (gkalpak) [inline packages/tsconfig-build-strict.json]: OOC, have you also tried `--noImplicitReturns`, `--noUnusedLocals`, `--noUnusedParameters`?
(I assume there will be a ton of violations :grin:)

**Review by gkalpak — COMMENTED** [review]
:tada:

File: packages/core/src/di/reflective_provider.ts [hunk 3]
```diff
@@ -180,10 +180,11 @@ export function mergeResolvedReflectiveProviders(
   return normalizedProvidersMap;
 }
 
-function _normalizeProviders(providers: Provider[], res: Provider[]): Provider[] {
+function _normalizeProviders(
+    providers: Provider[], res: NormalizedProvider[]): NormalizedProvider[] {
   providers.forEach(b => {
     if (b instanceof Type) {
-      res.push({provide: b, useClass: b});
+      res.push({ provide: b, useClass: b } as NormalizedProvider);
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/core/src/di/reflective_provider.ts]: Yes. I had the same feeling. `NormalizedProvider` seems to be incorrect since it extends from `TypeProvider`. Let's look into this in a separate PR/story.

File: packages/BUILD.bazel [hunk 4]
```diff
@@ -23,6 +23,12 @@ ts_config(
     deps = [":tsconfig-build.json"],
 )
 
+ts_config(
+    name = "tsconfig-build-strict",
+    src = "tsconfig-build-strict.json",
```
Comment (alexeagle) [inline packages/BUILD.bazel]: I'm trying to remove all the extra tsconfig files. What would be needed to avoid adding a new one? need a refactoring tool to automatically correct each `--strict` option as we enable it? We are quite good at doing that in google3 where we've always had a single tsconfig for the whole monorepo

File: packages/core/src/debug/debug_node.ts [hunk 5]
```diff
@@ -411,7 +411,7 @@ class DebugElement__POST_R3__ extends DebugNode__POST_R3__ implements DebugEleme
 
   queryAll(predicate: Predicate<DebugElement>): DebugElement[] {
     const matches: DebugElement[] = [];
-    _queryAllR3(this, predicate, matches, true);
+    _queryAllR3(this, predicate as Predicate<DebugNode>, matches, true);
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/core/src/debug/debug_node.ts]: Very good question. This is actually a bit odd because the predicate wants to be called with `DebugElement` instances, while the `_queryAllR3` function only *calls it* with `DebugNode`. 

Now with `strictFunctionTypes`.. this fails because technically something could use `DebugElement` specific logic in the predicate, while only a `DebugNode` got passed. See [here](https://www.typescriptlang.org/play/#src=interface%20Base%20%7B%0D%0A%20%20myProp%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20Derived%20extends%20Base%20%7B%0D%0A%20%20derivedProp%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Afunction%20test(cb%3A%20(c%3A%20Base)%20%3D%3E%20void)%20%7B%20%7D%0D%0A%0D%0Atest((a%3A%20Base)%20%3D%3E%20%7B%20%7D)%3B%0D%0Atest((a%3A%20Derived)%20%3D%3E%20%7B%7D)%3B)).

I'll add another commit that fixes this under-the-hood so that we don't need this "hacky" cast.

File: packages/compiler/src/render3/view/template.ts (hunk 1 above)
Comment (devversion) [reply packages/compiler/src/render3/view/template.ts]: I did that to make it clear that shared context variables should be of type `o.ReadVarExpr`, but looking at it again it seems weird. Removed.

File: packages/tsconfig-build-strict.json (hunk 2 above)
Comment (devversion) [reply packages/tsconfig-build-strict.json]: Nope not yet :smile: We should definitely look into that, but probably out of scope for the `--strict` change for now :grin:

File: packages/core/test/acceptance/ng_module_spec.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      class MyModule {
+      }
+
+      TestBed.configureTestingModule({imports: [MyModule]});
+
+      expect(() => TestBed.createComponent(TestCmp)).not.toThrow();
+    });
+  });
+
+});
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/core/test/acceptance/ng_module_spec.ts]: We actually only needed one call to `createComponent` to ensure that the test module is created. I'll either add a comment explaining this or going the safe and less-confusing way of calling `createComponent` for both components.

File: packages/core/src/util/decorators.ts [hunk 7]
```diff
@@ -49,10 +49,11 @@ export function makeDecorator<T>(
     {new (...args: any[]): any; (...args: any[]): any; (...args: any[]): (cls: any) => any;} {
   const metaCtor = makeMetadataCtor(props);
 
-  function DecoratorFactory(...args: any[]): (cls: Type<T>) => any {
+  function DecoratorFactory(this: void|typeof DecoratorFactory, ...args: any[]): (cls: Type<T>) =>
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/core/src/util/decorators.ts]: I like that more. done!

File: packages/core/src/debug/debug_node.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply packages/core/src/debug/debug_node.ts]: I didn't realize `_queryAllR3()` was eventually calling `predicate()` (with a `DebugNode`). Makes sense now.

File: packages/core/src/render3/di.ts [hunk 8]
```diff
@@ -557,7 +557,7 @@ export function getNodeInjectable(
     const saveLView = getLView();
     setTNodeAndViewData(tNode, lData);
     try {
-      value = lData[index] = factory.factory(null, tData, lData, tNode);
+      value = lData[index] = factory.factory(undefined, tData, lData, tNode);
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/core/src/render3/di.ts]: Very good question. The reason for this change is that the canonical factories are typed using the `FactoryFn` interface:

https://github.com/angular/angular/blob/0ee09cdd7e5f076355775322ae7d21de8f23c40a/packages/core/src/render3/interfaces/definition.ts#L45-L50

The `NodeInjectorFactory` currently calls these factories for directives with `null` as the first parameter. This is incorrect and could cause unexpected behavior. We could have solved this differently by wrapping the directive factory functions like that: `(_: null) => directiveFactory()`, but  this results in a few function declaration for each directive. Additionally it looks like the initial goal was to type the `NodeInjectorFactory#factory` equally to the `FactoryFn` (to avoid these wrappings).. AFAICT that's why we even pass around `null` (and now `undefined` :grin:)

File: packages/BUILD.bazel (hunk 4 above)
Comment (devversion) [reply packages/BUILD.bazel]: You mean something like a genrule to generate the modified tsconfig dynamically? Should we handle this outside of this PR? (seems like a general question of how we manage these configs)

File: packages/core/src/render3/styling/host_instructions_queue.ts [hunk 9]
```diff
@@ -61,7 +61,7 @@ function findNextInsertionIndex(buffer: HostInstructionsQueue, priority: number)
  * Iterates through the host instructions queue (if present within the provided
  * context) and executes each queued instruction entry.
  */
-export function flushQueue(context: StylingContext): void {
+export function flushQueue(this: any, context: StylingContext): void {
   const buffer = context[StylingIndex.HostInstructionsQueue];
```
Comment (petebacondarwin) [inline packages/core/src/render3/styling/host_instructions_queue.ts]: I see both `this: any` and `this: unknown`.
Is there a difference? If not then we should be consistent and just use one.

**Review by petebacondarwin — APPROVED** [review]

File: packages/core/src/render3/styling/host_instructions_queue.ts (hunk 9 above)
Comment (devversion) [reply packages/core/src/render3/styling/host_instructions_queue.ts]: Thanks Pete. Yes a few instances in this PR need to be switched to `unknown`. Note that also a few need to remain as `any` if the context is clear (but not type-able). e.g. `createDynamicClass` which uses `Object.create`.

File: packages/compiler/src/output/output_interpreter.ts [hunk 10]
```diff
@@ -79,7 +79,7 @@ function createDynamicClass(
 
   const ctorParamNames = _classStmt.constructorMethod.params.map(param => param.name);
   // Note: use `function` instead of arrow function to capture `this`
-  const ctor = function(...args: any[]) {
+  const ctor = function(this: any, ...args: any[]) {
```
Comment (petebacondarwin) [inline packages/compiler/src/output/output_interpreter.ts]: Actually I note that oin general using `any` hides potential bugs...
For example the `_ExecutionContext` constructor accepts an `instance` parameter which is `any` but on one occasion we are passing `null` here, then later in the code we access this value unguarded.  We could consider something like this, which at least makes the assertion that we believe `instance` is not `null` explicit:

```diff
diff --git a/packages/compiler/src/output/output_interpreter.ts b/packages/compiler/src/output/output_interpreter.ts
index 7f63f7bb6c..344fe37c06 100644
--- a/packages/compiler/src/output/output_interpreter.ts
+++ b/packages/compiler/src/output/output_interpreter.ts
@@ -38,8 +38,8 @@ class _ExecutionContext {
   exports: string[] = [];
 
   constructor(
-      public parent: _ExecutionContext|null, public instance: any, public className: string|null,
-      public vars: Map<string, any>) {}
+      public parent: _ExecutionContext|null, public instance: Object|null,
+      public className: string|null, public vars: Map<string, any>) {}
 
   createChildWihtLocalVars(): _ExecutionContext {
     return new _ExecutionContext(this, this.instance, this.className, new Map<string, any>());
@@ -79,9 +79,9 @@ function createDynamicClass(
 
   const ctorParamNames = _classStmt.constructorMethod.params.map(param => param.name);
   // Note: use `function` instead of arrow function to capture `this`
-  const ctor = function(...args: any[]) {
+  const ctor = function(this: Object, ...args: any[]) {
     const instanceCtx = new _ExecutionContext(_ctx, this, _classStmt.name, _ctx.vars);
-    _classStmt.fields.forEach((field) => { this[field.name] = undefined; });
+    _classStmt.fields.forEach((field) => { (this as any)[field.name] = undefined; });
     _executeFunctionStatements(
         ctorParamNames, args, _classStmt.constructorMethod.body, instanceCtx, _visitor);
   };
@@ -125,7 +125,7 @@ class StatementInterpreter implements o.StatementVisitor, o.ExpressionVisitor {
     if (ast.builtin != null) {
       switch (ast.builtin) {
         case o.BuiltinVar.Super:
-          return ctx.instance.__proto__;
+          return Object.getPrototypeOf(ctx.instance);
         case o.BuiltinVar.This:
           return ctx.instance;
         case o.BuiltinVar.CatchError:
@@ -188,7 +188,7 @@ class StatementInterpreter implements o.StatementVisitor, o.ExpressionVisitor {
     const args = this.visitAllExpressions(stmt.args, ctx);
     const fnExpr = stmt.fn;
     if (fnExpr instanceof o.ReadVarExpr && fnExpr.builtin === o.BuiltinVar.Super) {
-      ctx.instance.constructor.prototype.constructor.apply(ctx.instance, args);
+      ctx.instance !.constructor.prototype.constructor.apply(ctx.instance, args);
       return null;
     } else {
       const fn = stmt.fn.visitExpression(this, ctx);
```

Comment (devversion) [reply packages/compiler/src/output/output_interpreter.ts]: Totally agree. Even though this seems like a general loose type issue within the `output_interpreter` and is not necessarily in scope for this PR, I'm changing it to what you proposed in order to avoid that we need to type `this` with `any`.

File: packages/core/testing/src/r3_test_bed_compiler.ts [hunk 11]
```diff
@@ -508,20 +508,23 @@ export class R3TestBedCompiler {
...
-        Object.defineProperty(type, prop, descriptor);
-      }
-    });
+    this.initialNgDefs.forEach(
+        (value: [string, PropertyDescriptor | undefined], type: Type<any>) => {
+          const [prop, descriptor] = value;
+          if (!descriptor) {
+            // Delete operations are generally undesirable since they have performance implications
+            // on objects they were applied to. In this particular case, situations where this code
+            // is
```
Comment (alxhub) [inline packages/core/testing/src/r3_test_bed_compiler.ts]: Some formatting weirdness going on here :)

**Review by alxhub — APPROVED** [review]
😲 👏 

Amazing work, sir. Thank you!!

File: packages/core/testing/src/r3_test_bed_compiler.ts (hunk 11 above)
Comment (devversion) [reply packages/core/testing/src/r3_test_bed_compiler.ts]: Good catch :smile: Done!

File: packages/platform-browser/src/dom/debug/by.ts [hunk 12]
```diff
@@ -44,14 +44,14 @@ export class By {
...
+   * Match nodes that have the given directive present.
    *
    * @usageNotes
    * ### Example
    *
    * {@example platform-browser/dom/debug/ts/by/by.ts region='by_directive'}
    */
-  static directive(type: Type<any>): Predicate<DebugElement> {
-    return (debugElement) => debugElement.providerTokens !.indexOf(type) !== -1;
+  static directive(type: Type<any>): Predicate<DebugNode> {
```
Comment (devversion) [inline packages/platform-browser/src/dom/debug/by.ts]: These return type changes can seem like a breaking change, but `Predicate<DebugNode>` is also assignable to `Predicate<DebugElement>` it should not be breaking.

It just _fixes_ the type so that `debugEl.query(By.directive(..))` also works for consumer applications using `strictFunctionTypes`. Here is a playground demo: [see here](https://www.typescriptlang.org/play/?noImplicitAny=false&strictFunctionTypes=false&noImplicitReturns=false&alwaysStrict=false#code/LAKA9AVBoAQTBBGAjA9qgNgUwIYDsBaANxwwFcsATGAMzLwGMAXAS1TxlSKwCcYcYJclgA0MAA6oAzlJbIMATxgtG5SioDmMBuyZYAHk2V4aqHgFscrdrHg8sGnD3V4tTABZXBpCgHIpEtIs1hwq-BxOPDgKAHS2tjAAAuJk8iwMCOIstmCgKno8NDgMWDAACvbqDFZYADwAKgB8MADeMAAUQhQAXDD1AJS9aJi4eADcMAC+oHl4BUUlMAAiWMhkGgByqJSlLbAwB3jbWBWo4r1STDyaY6DTILPzxaUraxoAotjmWHMwBnp4SgBV7rLY7Vr7A5YDCnc4wS7XVy3ED3UB0RghfgMEriJhSEEfL4-JgVKjpGrtfS9UlVGq1AmfLDfOaNfqte7o5hsCLYrC4-GrUHHGnkvSU6mVUV1AlgrCs9kzEDFHF4hlEuYi6pi9p4XpqpnEtkAXmaVwo-WRyr5qsFhINGslWqw7R1ettsuNpp45uRoCt-JlwsdFNdy1tjOZTE9MDNWDZYDA8PcqDIGGoRRYGBgAHdgu54Vd0kY8Kmswx3FgGABrAJhHR4KRkb58HDicQYUXcmJ9dwsWsBAziSt6SgAQhg9lItBwmbI9j9vID7qDZKdLt1YbeHpgJpj3rjYyAA)

**Review by JiaLiPassion — APPROVED** [review]
Thanks!

**Review by IgorMinar — APPROVED** [review]
thanks Paul!!

**Comment by mhevery** [comment]
http://test/OCL:258796685:BASE:258796726:1563470812584:47aac9ff
